File: src/web/static/js/home/evidence_report.js
/* Live ScientificReport@1: detached display data, never ACK/result authority. */
(function (root) {
  "use strict";
  const trusted = new WeakSet();
  const trustedRows = new WeakSet(), trustedSources = new WeakSet();
  const descriptors = ["molecular_weight", "logp", "tpsa", "qed"];
  const positive = s => s === "available" || s === "partial";
  const states = ["available", "partial", "not_provided", "unavailable", "invalid", "ambiguous"];
  const reasons = ["none", "missing_source", "source_unavailable", "source_failed", "source_mismatch", "input_unverifiable",
    "alignment_missing", "row_missing", "invalid_value", "ambiguous_source", "stale_source", "candidate_not_displayed", "partial_source", "unsupported_binding"];
  const check = ok => { if (!ok) throw Error("Invalid scientific report"); };
  const exact = (o, fields) => check(o && !Array.isArray(o) && typeof o === "object" &&
    Object.keys(o).sort().join(" ") === fields.split(" ").sort().join(" "));
  const text = (v, max = 128, nullable = false) => check((nullable && v === null) || typeof v === "string" && v.trim() && v.length <= max);
  const digest = (v, nullable = false) => check((nullable && v === null) || typeof v === "string" && /^[a-f0-9]{64}$/.test(v));
  const count = (v, nullable = false) => check((nullable && v === null) || Number.isSafeInteger(v) && v >= 0);
  const unit = (v, nullable = false) => check((nullable && v === null) || typeof v === "number" && Number.isFinite(v) && v >= 0 && v <= 1);
  const list = (v, max) => check(Array.isArray(v) && v.length <= max);
  const missing = v => {list(v, 2); check(new Set(v).size === v.length && v.every(x => x === "admet" || x === "activity"));};
  const pairKey = (oid, cid) => JSON.stringify([oid, cid]);
  function sameReference(a, b) {
    return a && b && a.trace_id === b.trace_id && a.presentation_id === b.presentation_id && a.revision === b.revision &&
      JSON.stringify(a.ordered_keys) === JSON.stringify(b.ordered_keys);
  }
  function cloneBounded(value) {
    let nodes = 0, chars = 0;
    const active = new Set();
    function copy(v, depth) {
      check(++nodes <= 8192 && depth <= 12);
      if (v === null || typeof v === "boolean") return v;
      if (typeof v === "number") {check(Number.isFinite(v)); return v;}
      if (typeof v === "string") {
        chars += v.length; check(chars <= 131072);
        check(!/[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:^|[^\uD800-\uDBFF])[\uDC00-\uDFFF]/u.test(v));
        return v;
      }
      check(v && typeof v === "object" && !active.has(v));
      const array = Array.isArray(v);
      check(Object.getPrototypeOf(v) === (array ? Array.prototype : Object.prototype));
      const fields = Object.getOwnPropertyDescriptors(v), names = Reflect.ownKeys(fields);
      check(names.length <= 8192);
      if (array) check(names.length === v.length + 1 && v.length <= 8192);
      active.add(v);
      const out = array ? [] : {};
      for (const k of names) {
        if (array && k === "length") continue;
        check(typeof k === "string" && !["__proto__", "constructor", "prototype"].includes(k));
        const d = fields[k]; check("value" in d && d.enumerable);
        if (array) check(/^(0|[1-9][0-9]*)$/.test(k) && Number(k) < v.length);
        copy(k, depth + 1);
        out[k] = copy(d.value, depth + 1);
      }
      active.delete(v);
      return out;
    }
    const out = copy(value, 0);
    // UTF-8 byte budget; only detached exact builtins reach serialization.
    check(unescape(encodeURIComponent(JSON.stringify(out))).length <= 131072);
    return out;
  }
  function normalize(payload) {
    try {
      const p = cloneBounded(payload);
      exact(p, "type schema_version trace_id source_version projection_id run_status target sources generations collections property_rows steps ranking warnings omitted");
      check(p.type === "scientific_report" && p.schema_version === "1" && ["completed", "partial"].includes(p.run_status));
      text(p.trace_id); digest(p.source_version); digest(p.projection_id);
      exact(p.target, "label origin"); text(p.target.label, 128, true);
      check(p.target.origin === (p.target.label === null ? "not_provided" : "workflow_plan"));
      for (const [k, n] of Object.entries({sources:24, generations:8, collections:8, property_rows:32, steps:32, warnings:32})) list(p[k], n);
      const sources = new Map(), keys = new Set(), views = new Set();
      for (const s of p.sources) {
        exact(s, "observation_id step_id tool_name tool_version evidence_id status input_digest data_digest observation_digest output_digest_origin model_name model_version method backend_version");
        for (const k of ["observation_id", "step_id", "tool_name", "tool_version", "evidence_id"]) text(s[k]);
        for (const k of ["input_digest", "data_digest", "observation_digest"]) digest(s[k]);
        for (const k of ["model_name", "model_version", "method", "backend_version"]) text(s[k], 128, true);
        check(["succeeded", "partial"].includes(s.status) && ["recorded", "checkpoint_snapshot"].includes(s.output_digest_origin));
        check(!sources.has(s.observation_id)); sources.set(s.observation_id, s);
      }
      for (const c of p.collections) {
        exact(c, "reference generator_observation_id"); const r = c.reference;
        exact(r, "trace_id presentation_id revision ordered_keys");
        check(r.trace_id === p.trace_id); text(r.presentation_id); digest(r.revision);
        check(!views.has(r.presentation_id)); views.add(r.presentation_id);
        check(sources.get(c.generator_observation_id)?.tool_name === "llm_molecular_generator");
        list(r.ordered_keys, 32); check(r.ordered_keys.length > 0);
        for (const pair of r.ordered_keys) {
          list(pair, 2); check(pair.length === 2 && pair[0] === c.generator_observation_id); text(pair[1]);
          const key = pairKey(...pair); check(!keys.has(key)); keys.add(key);
        }
      }
      const generations = new Set();
      for (const g of p.generations) {
        const fields = "requested_count valid_count unique_count invalid_count duplicate_count displayed_count";
        exact(g, "source_observation_id status " + fields);
        check(["succeeded", "partial", "unknown"].includes(g.status));
        if (g.status === "unknown") check(g.source_observation_id === null && fields.split(" ").every(k => g[k] === null));
        else {
          check(sources.get(g.source_observation_id)?.tool_name === "llm_molecular_generator" && sources.get(g.source_observation_id).status === g.status);
          check(!generations.has(g.source_observation_id)); generations.add(g.source_observation_id);
          fields.split(" ").forEach(k => count(g[k]));
          check(g.displayed_count === [...keys].filter(k => JSON.parse(k)[0] === g.source_observation_id).length && g.displayed_count <= g.unique_count && g.unique_count <= g.valid_count);
        }
      }
      const rows = new Set();
      for (const r of p.property_rows) {
        exact(r, "generator_observation_id candidate_id canonical_smiles source_observation_id source_row_index row_digest state reason_code values");
        const key = pairKey(r.generator_observation_id, r.candidate_id);
        check(keys.has(key) && !rows.has(key)); rows.add(key); text(r.canonical_smiles, 8192);
        check(states.includes(r.state) && reasons.includes(r.reason_code)); exact(r.values, descriptors.join(" "));
        for (const [k, v] of Object.entries(r.values)) if (v !== null) {
          check(typeof v === "number" && Number.isFinite(v));
          check(k !== "molecular_weight" || v > 0); check(k !== "tpsa" || v >= 0); if (k === "qed") unit(v);
        }
        if (positive(r.state)) {
          check(sources.get(r.source_observation_id)?.tool_name === "property_calculator");
          count(r.source_row_index); digest(r.row_digest); check(Object.values(r.values).some(v => v !== null));
          if (r.state === "available") check(sources.get(r.source_observation_id).status === "succeeded" && Object.values(r.values).every(v => v !== null));
        } else check(Object.values(r.values).every(v => v === null) && r.source_observation_id === null && r.source_row_index === null && r.row_digest === null);
      }
      for (const s of p.steps) {
        exact(s, "step_id tool_name status source_observation_id reason_code message"); text(s.step_id); text(s.tool_name); text(s.message, 256);
        check(["succeeded", "partial", "failed", "rejected", "cancelled", "skipped", "unknown"].includes(s.status) && reasons.includes(s.reason_code));
        check(s.source_observation_id === null || sources.has(s.source_observation_id));
      }
      const rank = p.ranking;
      exact(rank, "state reason_code source_observation_id generator_observation_id requested_top_n ranked_candidate_count top_candidates unrankable_candidates");
      check(states.includes(rank.state) && reasons.includes(rank.reason_code)); list(rank.top_candidates, 32); list(rank.unrankable_candidates, 32);
      if (!positive(rank.state)) check(!rank.top_candidates.length && !rank.unrankable_candidates.length &&
        ["source_observation_id", "generator_observation_id", "requested_top_n", "ranked_candidate_count"].every(k => rank[k] === null));
      else {
        check(sources.get(rank.source_observation_id)?.tool_name === "candidate_ranker"); count(rank.requested_top_n); count(rank.ranked_candidate_count);
        check(rank.requested_top_n > 0 && rank.top_candidates.length <= Math.min(rank.requested_top_n, rank.ranked_candidate_count));
        const ranked = new Set();
        for (const r of [...rank.top_candidates, ...rank.unrankable_candidates]) {
          const key = pairKey(rank.generator_observation_id, r.candidate_id);
          check(keys.has(key) && !ranked.has(key)); ranked.add(key); text(r.canonical_smiles, 8192);
        }
        for (const r of rank.unrankable_candidates) {exact(r, "candidate_id canonical_smiles reason"); text(r.reason, 256);}
        for (const r of rank.top_candidates) {
          exact(r, "candidate_id canonical_smiles score missing_evidence ranking_evidence"); unit(r.score); missing(r.missing_evidence);
          const e = r.ranking_evidence; exact(e, "property_score admet_score activity_score weights_used missing_evidence");
          unit(e.property_score); unit(e.admet_score, true); unit(e.activity_score, true); missing(e.missing_evidence);
          check(JSON.stringify(e.missing_evidence) === JSON.stringify(r.missing_evidence)); exact(e.weights_used, "properties admet activity");
          for (const [k, v] of Object.entries(e.weights_used)) unit(v, k !== "properties");
        }
      }
      p.warnings.forEach(w => text(w, 256)); exact(p.omitted, "generations steps property_rows ranking_rows warnings"); Object.values(p.omitted).forEach(v => count(v));
      function freeze(v) {if (v && typeof v === "object") {Object.values(v).forEach(freeze); Object.freeze(v);}}
      freeze(p); trusted.add(p); p.property_rows.forEach(r => trustedRows.add(r)); p.sources.forEach(s => trustedSources.add(s)); return p;
    } catch (_) {return null;}
  }
  function createLifecycle() {
    let active = false, trace = null, value = null, conflicted = false;
    const clear = () => {active = false; trace = value = null; conflicted = false;};
    return {clear, startRequest() {clear(); active = true;},
      observeTrace(id) {if (!active || typeof id !== "string" || !id || id.length > 128) return false;
        if (trace && trace !== id) {value = null; conflicted = true; return false;} trace = id; return true;},
      enqueue(raw) {const p = normalize(raw); if (!active || conflicted || !p || p.trace_id !== trace) return false;
        if (value) {if (value.projection_id !== p.projection_id || JSON.stringify(value) !== JSON.stringify(p)) {value = null; conflicted = true;} return false;}
        value = p; return true;},
      take(id) {const p = active && !conflicted && id === trace ? value : null; clear(); return p;}};
  }
  const labels = {completed:"已完成", succeeded:"成功", partial:"部分完成／证据不完整", failed:"失败", rejected:"已拒绝", cancelled:"已取消", skipped:"已跳过", unknown:"未知",
    available:"已验证来源", not_provided:"未提供独立证据", unavailable:"不可用", invalid:"证据无效", ambiguous:"来源不唯一"};
  function line(parent, tag, text) {const el = document.createElement(tag); el.textContent = text; parent.appendChild(el); return el;}
  function renderReport(container, report) {
    if (!container || !trusted.has(report)) return {mounted:false};
    const box = document.createElement("section"); box.className = "scientific-evidence-report";
    box.style.cssText = "margin:16px 0;padding:16px;border:1px solid #cbd5e1;border-radius:8px;overflow-wrap:anywhere";
    line(box, "h3", `科研证据报告 · ${report.target.label || "靶点未提供"} · ${labels[report.run_status]}`);
    report.generations.forEach(g => line(box, "p", g.status === "unknown" ? "生成数量：未知" :
      `生成：请求 ${g.requested_count}，有效 ${g.valid_count}，去重后 ${g.unique_count}，无效 ${g.invalid_count}，重复 ${g.duplicate_count}，展示 ${g.displayed_count}`));
    report.steps.forEach(s => line(box, "p", `${s.tool_name}：${labels[s.status]}；${s.message}`));
    const rank = report.ranking;
    line(box, "h4", positive(rank.state) ? `实际 Top-${rank.requested_top_n}：展示 ${rank.top_candidates.length} / 可排序 ${rank.ranked_candidate_count}（${labels[rank.state]}）` : `候选排序：${labels[rank.state]}`);
    rank.top_candidates.forEach((r, i) => line(box, "p", `${i + 1}. ${r.candidate_id} · 优先级分数 ${r.score}；缺失证据：${r.missing_evidence.join("、") || "无"}`));
    rank.unrankable_candidates.forEach(r => line(box, "p", `${r.candidate_id}：不可排序（${r.reason}）`));
    report.sources.forEach(s => line(box, "p", `来源 ${s.tool_name}@${s.tool_version} · ${s.model_name || s.method || "模型未提供"} · 版本 ${s.model_version || s.backend_version || "未提供"} · ${s.evidence_id} · ${s.output_digest_origin}`));
    report.warnings.forEach(w => line(box, "p", w));
    if (Object.values(report.omitted).some(v => v > 0)) line(box, "p", `展示截断：${JSON.stringify(report.omitted)}`);
    container.appendChild(box); return {mounted:true};
  }
  function findPropertyRow(report, reference, generatorId, candidate) {
    if (!trusted.has(report) || !report.collections.some(c => c.generator_observation_id === generatorId && sameReference(c.reference, reference))) return null;
    return report.property_rows.find(r => r.generator_observation_id === generatorId && r.candidate_id === candidate.candidate_id && r.canonical_smiles === candidate.canonical_smiles) || null;
  }
  function renderProperties(container, row, source) {
    if (!container || !trustedRows.has(row) || !trustedSources.has(source) || !positive(row.state) || source.observation_id !== row.source_observation_id || source.tool_name !== "property_calculator") return false;
    const box = document.createElement("div");
    line(box, "strong", `独立 RDKit 描述符 · ${labels[row.state]}`);
    const names = ["分子量", "LogP", "TPSA", "QED"];
    descriptors.forEach((k, i) => line(box, "div", `${names[i]}：${row.values[k] === null ? "未提供" : row.values[k]}`));
    line(box, "small", `来源 ${source.evidence_id} · 行 ${row.source_row_index} · 非实验验证`);
    container.replaceChildren(box); return true;
  }
  root.HomeEvidenceReport = {normalize, createLifecycle, renderReport, findPropertyRow, renderProperties};
})(window);

File: src/agent/presentation/evidence_report.py
"""Live static-workflow projection. Stored observations, not model prose, are authority.

No tool execution, result repair, ledger writes, or property recalculation here.
Temporary ToolResult/ledger instances below verify identities on detached copies.
"""
import json
from hashlib import sha256

from src.agent.contracts.scientific_references import reference_json
from src.agent.contracts.scientific_report import DESCRIPTORS, descriptor_values, validate_report
from src.agent.evidence.ledger import EvidenceLedger
from src.agent.orchestrators import WorkflowOrchestrator
from src.agent.persistence.scientific_references import _candidates
from src.agent.persistence.redaction import sanitize_sensitive_text
from src.agent.planning.bindings import BindingResolver
from src.agent.tooling.analysis_contract import _AnalysisRawOutput


def _check(condition):
    if not condition:
        raise ValueError("report source proof unavailable")


def _safe(text, limit=128):
    clean, changed = sanitize_sensitive_text(text, max_chars=limit)
    return clean if clean and not changed else None


def _accepted(cp, run, sequence, ledger):
    raw = json.loads(cp["output_json"])
    reference_json(raw)
    _check(cp["trace_id"] == run["trace_id"] and cp["workflow_version"] == run["workflow_version"])
    _check(cp["status"] in {"succeeded", "partial"} and cp["error_json"] is None
           and raw["success"] is True and raw["status"] == cp["status"] and raw.get("error") is None)
    expected = {**raw, "tool_name": cp["tool_name"]}
    matches = [o for o in sequence if o.get("quality", {}).get("step_id") == cp["step_id"]]
    _check(len(matches) == 1)
    observed = {k: v for k, v in matches[0].items() if k != "step_id"}
    _check(reference_json(observed) == reference_json(expected))
    _check(raw["quality"]["step_id"] == cp["step_id"])
    _check(raw["quality"].get("demo_mode", False) is False
           and raw["quality"].get("fallback_used", False) is False)
    result = WorkflowOrchestrator._result_from_checkpoint(cp["tool_name"], {"output": raw, "status": cp["status"]})
    p = result.provenance
    _check(p is not None and not p.demo_mode and not p.fallback_used
           and p.tool_name == cp["tool_name"] and p.tool_version == cp["tool_version"]
           and p.input_digest == cp["input_hash"])
    digest = EvidenceLedger.output_digest(raw["data"])
    _check(p.output_digest is None or p.output_digest == digest)
    verifier = EvidenceLedger(run["trace_id"])
    eid = verifier.register_tool_result(cp["step_id"], cp["input_hash"], result)
    _check(eid == raw["quality"]["evidence_id"])
    if ledger is not None:
        _check(type(ledger) is list)
        records = [e for e in ledger if e.get("evidence_id") == eid]
        _check(len(records) == 1 and reference_json(records[0]) == reference_json(verifier.get(eid)))
    if cp["tool_name"] == "llm_molecular_generator":
        _candidates(cp, run["trace_id"])
    elif cp["tool_name"] == "property_calculator":
        _AnalysisRawOutput.model_validate(raw, context={"tool_name": "property_calculator"})
    source = dict(observation_id=cp["id"], step_id=cp["step_id"], tool_name=cp["tool_name"],
        tool_version=cp["tool_version"], evidence_id=eid, status=cp["status"],
        input_digest=cp["input_hash"], data_digest=digest,
        observation_digest=sha256(reference_json(expected).encode("utf-8")).hexdigest(),
        output_digest_origin="recorded" if p.output_digest else "checkpoint_snapshot",
        model_name=_safe(p.model_name), model_version=_safe(p.model_version),
        method="rdkit_descriptors" if cp["tool_name"] == "property_calculator" else None,
        backend_version=None)
    _check(all(_safe(source[k]) == source[k] for k in
               ("observation_id", "step_id", "tool_name", "tool_version", "evidence_id")))
    return {"checkpoint": cp, "raw": raw, "source": source, "metadata": json.loads(cp["metadata_json"])}


def _property_rows(prop, generator, run):
    cp, raw = prop["checkpoint"], prop["raw"]
    key = generator["raw"]["quality"]["output_key"]
    _check(prop["metadata"].get("candidate_source") == key)
    data = generator["raw"]["data"]
    bound = BindingResolver().resolve(f"$.outputs.{key}", "smiles_text", {}, {key: data})
    _check(WorkflowOrchestrator._input_hash(bound) == cp["input_hash"])
    from rdkit import Chem
    candidates = {c["candidate_id"]: c for c in data["candidates"]}
    rows = {}
    for index, row in enumerate(raw["data"]):
        cid = row.get("candidate_id")
        _check(cid in candidates and cid not in rows)
        molecule = Chem.MolFromSmiles(row["smiles"])
        _check(molecule is not None and Chem.MolToSmiles(molecule, canonical=True) == candidates[cid]["canonical_smiles"])
        values = {k: row["properties"].get(k) for k in DESCRIPTORS}
        descriptor_values(values)
        rows[cid] = (index, row, values)
    alignment = raw["quality"].get("candidate_alignment")
    _check(type(alignment) is dict and set(alignment) == {
        "source_count", "aligned_count", "missing_candidate_ids", "discarded_count"})
    _check(type(alignment["source_count"]) is int and alignment["source_count"] == len(candidates)
           and type(alignment["aligned_count"]) is int and alignment["aligned_count"] == len(rows)
           and alignment["missing_candidate_ids"] == [c for c in candidates if c not in rows]
           and type(alignment["discarded_count"]) is int and alignment["discarded_count"] >= 0)
    _check(raw["status"] == "partial" or (not alignment["missing_candidate_ids"] and alignment["discarded_count"] == 0))
    return rows


def _empty_ranking(reason="missing_source"):
    return dict(state="not_provided", reason_code=reason, source_observation_id=None,
        generator_observation_id=None, requested_top_n=None, ranked_candidate_count=None,
        top_candidates=[], unrankable_candidates=[])


def _ranking(accepted, outputs, generators, run, displayed):
    ranks = [s for s in accepted.values() if s["source"]["tool_name"] == "candidate_ranker"]
    if not ranks:
        return _empty_ranking()
    if len(ranks) != 1:
        return _empty_ranking("ambiguous_source")
    rank = ranks[0]
    try:
        meta = rank["metadata"]
        _check(meta.get("workflow_output_keys") == ["molecules", "properties", "admet", "activity"]
               and meta.get("workflow_optional_output_keys") == ["admet", "activity"]
               and meta.get("workflow_metadata_keys") == ["docking_top_n"])
        generators = [g for g in generators if g["raw"]["quality"].get("output_key") == "molecules"]
        _check(len(generators) == 1)
        generator = generators[0]
        # Refuse ranking unless the required properties are independently aligned too.
        props = [s for s in accepted.values() if s["raw"]["quality"].get("output_key") == "properties"]
        _check(len(props) == 1 and props[0]["source"]["tool_name"] == "property_calculator")
        _property_rows(props[0], generator, run)
        request = {"query": run["query"], "metadata": {"docking_top_n": meta["docking_top_n"]}}
        bound = BindingResolver().resolve("$.workflow", "identity", request, outputs,
            workflow_output_keys=meta["workflow_output_keys"],
            workflow_optional_output_keys=meta["workflow_optional_output_keys"],
            workflow_metadata_keys=meta["workflow_metadata_keys"])
        _check(WorkflowOrchestrator._input_hash(bound) == rank["checkpoint"]["input_hash"])
        raw = rank["raw"]["data"]
        _check(rank["raw"]["quality"].get("output_contract") == "CandidateRanking@1")
        ranked = raw["ranked_candidates"]
        _check(type(raw["requested_top_n"]) is int and raw["requested_top_n"] == meta["docking_top_n"]
               and type(raw["ranked_candidate_count"]) is int and raw["ranked_candidate_count"] == len(ranked)
               and raw["top_candidates"] == ranked[:raw["requested_top_n"]])
        canonical = {c["candidate_id"]: c["canonical_smiles"] for c in generator["raw"]["data"]["candidates"]}
        seen = set()
        for row in ranked + raw["unrankable_candidates"]:
            _check(row["candidate_id"] not in seen and canonical.get(row["candidate_id"]) == row["canonical_smiles"])
            seen.add(row["candidate_id"])
        _check(seen == set(canonical))
        oid = generator["checkpoint"]["id"]
        # Never filter/re-sort Top-N into a different positive result.
        _check(all((oid, r["candidate_id"]) in displayed for r in raw["top_candidates"] + raw["unrankable_candidates"]))
        top = [{k: r[k] for k in ("candidate_id", "canonical_smiles", "score", "missing_evidence", "ranking_evidence")}
               for r in raw["top_candidates"][:32]]
        return dict(state="partial" if rank["raw"]["status"] == "partial" or any(r["missing_evidence"] for r in top) else "available",
            reason_code="partial_source" if rank["raw"]["status"] == "partial" or any(r["missing_evidence"] for r in top) else "none",
            source_observation_id=rank["checkpoint"]["id"], generator_observation_id=oid,
            requested_top_n=raw["requested_top_n"], ranked_candidate_count=len(ranked),
            top_candidates=top, unrankable_candidates=raw["unrankable_candidates"][:32])
    except (ValueError, TypeError, KeyError, AttributeError):
        return _empty_ranking("input_unverifiable")


def build_evidence_report(snapshot, execution, candidate_events):
    """Fail closed; callers may keep all legacy frames when this returns None."""
    try:
        if snapshot is None:
            return None
        run = snapshot["run"]
        _check(execution.get("trace_id") == run["trace_id"] and run["skill_name"] == "target_driven_design")
        status = "partial" if run["status"] == "partial" else "completed"
        _check(execution.get("status") == status)
        plan = execution.get("workflow_plan", {})
        _check(plan.get("workflow_name") == "target_driven_design")
        sequence = execution.get("tool_result_sequence")
        reference_json(sequence)
        _check(type(sequence) is list and len(sequence) <= 256)
        ledger = execution.get("metadata", {}).get("evidence_ledger")
        if ledger is None and execution.get("agent_result") is not None:
            ledger = execution["agent_result"].metadata.get("evidence_ledger")
        accepted, failures = {}, {}
        for cp in snapshot["latest"].values():
            try:
                accepted[cp["id"]] = _accepted(cp, run, sequence, ledger)
            except (ValueError, TypeError, KeyError, AttributeError):
                failures[cp["id"]] = "source_failed" if cp["status"] not in {"succeeded", "partial"} else "source_mismatch"
        _check(len(accepted) <= 24)
        # No output-key collapse: duplicate providers invalidate that binding.
        keys = {}
        for s in accepted.values():
            key = s["raw"]["quality"].get("output_key")
            if type(key) is str:
                keys.setdefault(key, []).append(s)
        outputs = {k: items[0]["raw"]["data"] for k, items in keys.items() if len(items) == 1}
        collections, generations, rows, generators = [], [], [], []
        displayed = set()
        for view in snapshot["presentations"]:
            ref = {"trace_id": run["trace_id"], "presentation_id": view["presentation_id"],
                "revision": view["revision"], "ordered_keys": [[r["observation_id"], r["candidate"]["candidate_id"]]
                                                         for r in view["ordered_candidates"]]}
            events = [e for e in candidate_events if e.get("reference") == ref]
            _check(len(events) == 1 and events[0]["trace_id"] == run["trace_id"])
            oids = {r["observation_id"] for r in view["ordered_candidates"]}
            _check(len(oids) == 1)
            oid = next(iter(oids))
            generator = accepted[oid]
            _check(generator["source"]["tool_name"] == "llm_molecular_generator")
            data = generator["raw"]["data"]
            _check(reference_json(events[0]["candidate_set"]) == reference_json(data))
            _check(all(r["candidate"] in data["candidates"] for r in view["ordered_candidates"]))
            generators.append(generator)
            displayed.update(map(tuple, ref["ordered_keys"]))
            collections.append({"reference": ref, "generator_observation_id": oid})
            generations.append(dict(source_observation_id=oid, status=generator["raw"]["status"],
                **{k: data[k] for k in ("requested_count", "valid_count", "unique_count", "invalid_count", "duplicate_count")},
                displayed_count=len(ref["ordered_keys"])))
            key = generator["raw"]["quality"].get("output_key")
            dynamic = any(s["raw"]["quality"].get("request_input_digest") is not None for s in accepted.values())
            props = [s for s in accepted.values() if s["source"]["tool_name"] == "property_calculator"
                     and s["metadata"].get("candidate_source") == key]
            reason, mapped, prop = "missing_source", {}, None
            if dynamic:
                reason = "unsupported_binding"
            elif len(props) > 1 or len(keys.get(key, [])) != 1:
                reason = "ambiguous_source"
            elif props:
                try:
                    mapped = _property_rows(props[0], generator, run)
                    prop = props[0]
                except (ValueError, TypeError, KeyError, AttributeError):
                    reason = "input_unverifiable"
            elif any(cp["tool_name"] == "property_calculator" for cp in snapshot["latest"].values()):
                reason = "source_unavailable"
            for item in view["ordered_candidates"]:
                candidate = item["candidate"]
                row = dict(generator_observation_id=oid, candidate_id=candidate["candidate_id"],
                    canonical_smiles=candidate["canonical_smiles"], source_observation_id=None,
                    source_row_index=None, row_digest=None, state="not_provided", reason_code=reason,
                    values=dict.fromkeys(DESCRIPTORS))
                if prop is not None and candidate["candidate_id"] in mapped:
                    index, raw, values = mapped[candidate["candidate_id"]]
                    partial = prop["raw"]["status"] == "partial" or any(v is None for v in values.values())
                    row.update(source_observation_id=prop["checkpoint"]["id"], source_row_index=index,
                        row_digest=EvidenceLedger.output_digest(raw), values=values,
                        state="partial" if partial else "available", reason_code="partial_source" if partial else "none")
                elif prop is not None:
                    row["reason_code"] = "row_missing"
                rows.append(row)
        _check(collections and len(displayed) <= 32)
        steps = []
        for cp in snapshot["latest"].values():
            source = accepted.get(cp["id"])
            steps.append(dict(step_id=cp["step_id"], tool_name=cp["tool_name"],
                status=cp["status"] if source or cp["status"] in {"failed", "rejected", "cancelled"} else "unknown",
                source_observation_id=cp["id"] if source else None,
                reason_code="none" if source else failures.get(cp["id"], "source_unavailable"),
                message="工具执行记录；不等同于实验验证" if source else "未取得可验证的成功工具证据"))
        # Retain skipped tool outcomes without inventing successful observations.
        for skipped in execution.get("metadata", {}).get("skipped_steps", []):
            steps.append(dict(step_id=skipped["step_id"], tool_name=skipped.get("tool_name") or "未提供", status="skipped",
                source_observation_id=None, reason_code="source_unavailable", message="步骤已跳过"))
        target = _safe(plan.get("metadata", {}).get("target_hint"))
        warnings = ["计算描述符及候选优先级不代表实验验证、活性、疗效或对接结合能。",
                    "性质数字仅对本次实时证据展示有效；刷新后不恢复数字。"]
        for warning in execution.get("warnings", []):
            text = _safe(warning, 256)
            if text:
                warnings.append(text)
        ranking = _empty_ranking("unsupported_binding") if dynamic else _ranking(accepted, outputs, generators, run, displayed)
        report = dict(type="scientific_report", schema_version="1", trace_id=run["trace_id"],
            source_version=snapshot["source_version"], run_status=status,
            target={"label": target, "origin": "workflow_plan" if target else "not_provided"},
            sources=[s["source"] for s in accepted.values()], generations=generations,
            collections=collections, property_rows=rows, steps=steps[:32], ranking=ranking, warnings=warnings[:32],
            omitted=dict(generations=0, steps=max(0, len(steps)-32), property_rows=0,
                         ranking_rows=0, warnings=max(0, len(warnings)-32)))
        report["projection_id"] = sha256(reference_json(report).encode("utf-8")).hexdigest()
        return validate_report(report)
    except (ValueError, TypeError, KeyError, AttributeError, RecursionError):
        return None
